### PR TITLE
Various os_capacity fixes

### PR DESCRIPTION
**Pull request by assumptionsandg**
- Various typo fixes.
- Renamed `os_exporter` to `os_capacity`.
- Renamed `70-oscapacity.yml.j2` to `70-oscapacity.yml` to ensure correct templating.
- Added `70-oscapacity.yml` to .yamlint ignore list.
- Conditional flag `stackhpc_enable_os_capacity` added to prevent HAProxy config and Prometheus scrape targets being copied in before manual deployment.
- Conditional flag `stackhpc_os_capacity_openstack_verify` added where certificate verification can be overridden.
- Fixed interval errors

### CODE REVIEW

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (priteau) [inline etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: capacity

File: etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
Comment (priteau) [inline etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg]: Please add new line

File: doc/source/configuration/monitoring.rst
Comment (priteau) [inline doc/source/configuration/monitoring.rst]: conditional

**Review by priteau — CHANGES_REQUESTED** [review]
Existing docs say:

> StackHPC Kayobe Config includes this exporter by default

Should this be changed? Please also fix the typo: `avaliable`

File: doc/source/configuration/monitoring.rst
Comment (holliefae) [reply doc/source/configuration/monitoring.rst]: Done

File: etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
Comment (holliefae) [reply etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg]: Done

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (holliefae) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: Done

**Comment by assumptionsandg** [comment]
> Existing docs say:
> 
> > StackHPC Kayobe Config includes this exporter by default
> 
> Should this be changed? Please also fix the typo: `avaliable`

I've made an update to the wording to show deployment needs to be triggered manually, hopefully that's better.

File: etc/kayobe/stackhpc-monitoring.yml
Comment (JohnGarbutt) [inline etc/kayobe/stackhpc-monitoring.yml:17]: cool, we are off by default now, thank you.

Comment (holliefae) [reply etc/kayobe/stackhpc-monitoring.yml:17]: Right now the deployment requires manual intervention, it would be cool to get it automated in the future.

**Review by mnasiadka — DISMISSED** [review]

File: .yamllint
Comment (markgoddard) [inline .yamllint]: This file was renamed

File: etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml
Comment (markgoddard) [inline etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml:1]: Please see https://github.com/stackhpc/stackhpc-kayobe-config/pull/678#discussion_r1356595260

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml
Comment (holliefae) [reply etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml:1]: Done

File: .yamllint
Comment (holliefae) [reply .yamllint]: This was due to `70.oscapacity.yml.j2` not being templated as `70-oscapacity.yml` during deployment which caused Prometheus to not pick up the scrape job. I removed the .j2 extension and added it to the yamllint to stop ci failing it for the raw tags.

Comment (markgoddard) [reply .yamllint:23]:
```suggestion
  *etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-os-capacity.yml
```

Comment (markgoddard) [reply .yamllint]: hmm, I thought it had been renamed to 70-os-capacity.yml, but looks like I was wrong.

File: etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml
Comment (markgoddard) [inline etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml]: Unsure if this will render properly with indented jinja conditions
```suggestion
{% if kolla_enable_tls_internal | bool %}
    scheme: https
{% endif %}
```

Comment (holliefae) [reply etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml]: Dropped this commit as the os_capacity exporter does not expose a TLS endpoint.

**Review by Alex-Welsh — DISMISSED** [review]
LGTM

**Comment by assumptionsandg** [comment]
- Minor dashboard updates for graph intervals.
- Dashboard overhaul for OpenStack Project Metrics now including project and instance utilization

File: .yamllint
Comment (markgoddard) [inline .yamllint]: Why the leading asterisk?

File: etc/kayobe/ansible/deploy-os-capacity-exporter.yml
Comment (markgoddard) [inline etc/kayobe/ansible/deploy-os-capacity-exporter.yml]: Does it still need the mount?

Comment (holliefae) [reply etc/kayobe/ansible/deploy-os-capacity-exporter.yml]: Commited this by mistake.

File: .yamllint
Comment (holliefae) [reply .yamllint]: I was following convention from other entries in the file. Removing it doesn't seem to break CI.

File: etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml
Comment (markgoddard) [reply etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml:1]: Not done

File: etc/kayobe/kolla/config/grafana/dashboards/openstack/grafana_cloud_dashboard.json
Comment (dougszumski) [inline etc/kayobe/kolla/config/grafana/dashboards/openstack/grafana_cloud_dashboard.json]: I think we want to preserve the original line here. Please see this for context:

https://github.com/stackhpc/stackhpc-kayobe-config/commit/9dcc4ec836eefa4f2f797a31175cc860248a7107

File: etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml
Comment (dougszumski) [reply etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml:1]: Please can we fix this as Mark suggests @assumptionsandg? The naming is a bit confusing. For reference there are both `kolla_enable_tls_internal` and `kolla_enable_tls_backend` flags. The former terminates the TLS connection at the internal VIP, so the backend doesn't need to support TLS in that case.

**Review by dougszumski — CHANGES_REQUESTED** [review]
Thanks @assumptionsandg , just a couple of small changes needed.

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (markgoddard) [inline etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: It should be possible to specify cacert instead, in the same way as is done in ansible/roles/prometheus/templates/clouds.yml.j2

File: etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
Comment (markgoddard) [inline etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg]: This is the wrong place for this config. It needs to go in prometheus scrape config, etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml.

File: etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml
Comment (markgoddard) [inline etc/kayobe/kolla/config/prometheus/prometheus.yml.d/70-oscapacity.yml:10]: Please use `kolla_internal_fqdn` here instead of `kolla_internal_vip_address`.

**Review by markgoddard — DISMISSED** [review]

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (dougszumski) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: Please could you comment on how this was resolved @assumptionsandg? Thanks!

**Review by dougszumski — DISMISSED** [review]

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (dougszumski) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: I'm happy for this to be followed up in  another PR

File: doc/source/configuration/monitoring.rst
Comment (priteau) [inline doc/source/configuration/monitoring.rst]: Can you fix the `avaliable` typo while here?

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (priteau) [inline etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2:12]: Is there no way we can copy or bind mount the CA certificate file into the container?

**Review by priteau — DISMISSED** [review]
I left some minor comments.

File: etc/kayobe/stackhpc-monitoring.yml
Comment (jovial) [inline etc/kayobe/stackhpc-monitoring.yml]:
```suggestion
```
Remove extra section separator

**Review by jovial — COMMENTED** [review]
Tested and it works for me, thanks. Perhaps we could make copying the ca into the container a separate work item? It would be nice to get theses fixes in ASAP.

EDIT: When I said it was working, I had only tried a manual curl, see: https://github.com/stackhpc/stackhpc-kayobe-config/pull/697/files#r1408057734

File: etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
Comment (jovial) [inline etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg]: Just noticed this file is still callled os_exporter

**Comment by priteau** [comment]
> Tested and it works for me, thanks. Perhaps we could make copying the ca into the container a separate work item? It would be nice to get theses fixes in ASAP.

I agree, this can be added later.

File: etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
Comment (jovial) [reply etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg]: Also, I found prometheus could not scrape the exporter when internal TLS is enabled without:
```
diff --git a/etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg b/etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
index a0469006..c0bc2505 100644
--- a/etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
+++ b/etc/kayobe/kolla/config/haproxy/services.d/os_exporter.cfg
@@ -6,7 +6,11 @@ frontend os_capacity_frontend
     option httplog
     option forwardfor
     http-request set-header X-Forwarded-Proto https if { ssl_fc }
+    {% if kolla_enable_tls_internal | bool %}
+    bind {{ kolla_internal_vip_address }}:9000 ssl crt /etc/haproxy/haproxy-internal.pem
+    {% else %}
     bind {{ kolla_internal_vip_address }}:9000
+    {% endif %}
     default_backend os_capacity_backend
 
 backend os_capacity_backend
````

**Comment by jovial** [comment]
One other thing I've noticed on antelope is that some of the panels are empty e.g
image
but there is data in prometheus:
image
Recreating the panel worked, but I couldn't obviously figure out why it was returning no data.

**Comment by assumptionsandg** [comment]
> One other thing I've noticed on antelope is that some of the panels are empty e.g image but there is data in prometheus: image Recreating the panel worked, but I couldn't obviously figure out why it was returning no data.



> One other thing I've noticed on antelope is that some of the panels are empty e.g image but there is data in prometheus: image Recreating the panel worked, but I couldn't obviously figure out why it was returning no data.

Is this specifically a problem you've observed on Antelope that you haven't seen on older versions? 

I wonder if this is caused by the scrape interval being quite long, maybe Grafana is set to an interval by default where data was not available?

**Comment by assumptionsandg** [comment]
I haven't been able to replicate the empty Grafana panels in Yoga where the patch is intended.

 I was able to replicate the issues in Antelope when the interval is set to 30m, haven't found the underlying reason on why Yoga works and Antelope does not in this case.

**Comment by priteau** [comment]
Can you please rebase on top of `stackhpc/yoga`?

**Comment by assumptionsandg** [comment]
> One other thing I've noticed on antelope is that some of the panels are empty e.g image but there is data in prometheus: image Recreating the panel worked, but I couldn't obviously figure out why it was returning no data.

This issue is fixed in [978db2f](https://github.com/stackhpc/stackhpc-kayobe-config/pull/697/commits/978db2f82f443abd123331ebeb131f2d1d7996e8) Resolved by adding `avg_over_time` to the prometheus queries and reducing interval time to 10m.

File: etc/kayobe/kolla/config/haproxy/services.d/os_capacity.cfg
Comment (JohnGarbutt) [inline etc/kayobe/kolla/config/haproxy/services.d/os_capacity.cfg]: This part of the change needs to be in a PR to stable/2023.1 branch I think.

Comment (holliefae) [reply etc/kayobe/kolla/config/haproxy/services.d/os_capacity.cfg]: Removed, will readd for antelope.

File: doc/source/configuration/monitoring.rst
Comment (technowhizz) [inline doc/source/configuration/monitoring.rst]:
```suggestion
HAProxy and Prometheus configuration to be templated to controllers during deployment.
```

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (technowhizz) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: @assumptionsandg Is this going to be done in another PR or this one?

**Review by technowhizz — DISMISSED** [review]

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (holliefae) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: @technowhizz Probably a different pull request for this, most likely will bind mount self-signed certificates in to the container in the future to avoid disabling verification.

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (JohnGarbutt) [inline etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: I think this is the wrong variable here... we should be using the external endpoint in most case, that has a valid cert.

Lets just add verify false as an available workaround via a new flag like os_capacity_openstack_verify?

We should do what the other said about binding mounting the certificate in here, and I think there must be an existing example of that somewhere :/ But I would rather we didn't block on that here.

**Review by JohnGarbutt — DISMISSED** [review]

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (holliefae) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: Implemented in 8490b741c7ee3e43d20f2695db3ad6be53ccee86

Comment (holliefae) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: Certificate verification in #896

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (holliefae) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: Implemented in #896

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (holliefae) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2:12]: Moving TLS verification to #896

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (cityofships) [inline etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: Is this a secret?

Comment (holliefae) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: Doesn't really need to be since it's just the keystone auth url, this config is deprecated in #829 though.

File: doc/source/configuration/monitoring.rst
Comment (priteau) [inline doc/source/configuration/monitoring.rst]: s/You/you/

File: etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2
Comment (priteau) [reply etc/kayobe/ansible/templates/os_capacity-clouds.yml.j2]: So can you change it to use a Kolla Ansible variable?

File: doc/source/configuration/monitoring.rst
Comment (priteau) [inline doc/source/configuration/monitoring.rst]: In many environments, HAProxy and Prometheus are not deployed on controllers, so this could be confusing. You can just remove `to controllers`. Could you also explicitly state that both HAProxy and Prometheus must be reconfigured?

**Review by priteau — CHANGES_REQUESTED** [review]

File: doc/source/configuration/monitoring.rst
Comment (priteau) [inline doc/source/configuration/monitoring.rst:181]: Further down you can also change the playbook command to:

```
kayobe playbook run $KAYOBE_CONFIG_PATH/ansible/deploy-os-capacity-exporter.yml
```

This way it doesn't assume that you are in a specific directory.

**Review by priteau — CHANGES_REQUESTED** [review]
I also found another issue: `RegionOne` is hardcoded…

**Comment by priteau** [comment]
The URL of the os_capacity image is hardcoded to `ghcr.io/stackhpc/os-capacity:master`. We should allow it to be synchronised in Pulp like the other images we use.

**Review by priteau — CHANGES_REQUESTED** [review]
We also recently found out that reconfiguring HAProxy should be using the `loadbalancer` tag.

**Comment by priteau** [comment]
There is also a conflict caused by the rename of the HAProxy file: deployments which already have deployed `os_exporter.cfg` will get the following HAProxy error:

```
[ALERT]    (42) : Parsing [/etc/haproxy/services.d/os_capacity.cfg:2]: frontend 'os_capacity_frontend' has the same name as frontend 'os_capacity_frontend' declared at /etc/haproxy/services.d/os_exporter.cfg:2.
[ALERT]    (42) : Parsing [/etc/haproxy/services.d/os_capacity.cfg:11]: backend 'os_capacity_backend' has the same name as backend 'os_capacity_backend' declared at /etc/haproxy/services.d/os_exporter.cfg:11.
[ALERT]    (42) : Error(s) found in configuration file : /etc/haproxy/services.d/os_capacity.cfg
```

Operators need to be aware that they must delete `/etc/kolla/haproxy/services.d/os_exporter.cfg` first.

**Comment by mnasiadka** [comment]
> There is also a conflict caused by the rename of the HAProxy file: deployments which already have deployed `os_exporter.cfg` will get the following HAProxy error:
> 
> ```
> [ALERT]    (42) : Parsing [/etc/haproxy/services.d/os_capacity.cfg:2]: frontend 'os_capacity_frontend' has the same name as frontend 'os_capacity_frontend' declared at /etc/haproxy/services.d/os_exporter.cfg:2.
> [ALERT]    (42) : Parsing [/etc/haproxy/services.d/os_capacity.cfg:11]: backend 'os_capacity_backend' has the same name as backend 'os_capacity_backend' declared at /etc/haproxy/services.d/os_exporter.cfg:11.
> [ALERT]    (42) : Error(s) found in configuration file : /etc/haproxy/services.d/os_capacity.cfg
> ```
> 
> Operators need to be aware that they must delete `/etc/kolla/haproxy/services.d/os_exporter.cfg` first.

I guess it would make sense to delete that in the deployment playbook

**Comment by priteau** [comment]
> > There is also a conflict caused by the rename of the HAProxy file: deployments which already have deployed `os_exporter.cfg` will get the following HAProxy error:
> > ```
> > [ALERT]    (42) : Parsing [/etc/haproxy/services.d/os_capacity.cfg:2]: frontend 'os_capacity_frontend' has the same name as frontend 'os_capacity_frontend' declared at /etc/haproxy/services.d/os_exporter.cfg:2.
> > [ALERT]    (42) : Parsing [/etc/haproxy/services.d/os_capacity.cfg:11]: backend 'os_capacity_backend' has the same name as backend 'os_capacity_backend' declared at /etc/haproxy/services.d/os_exporter.cfg:11.
> > [ALERT]    (42) : Error(s) found in configuration file : /etc/haproxy/services.d/os_capacity.cfg
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Operators need to be aware that they must delete `/etc/kolla/haproxy/services.d/os_exporter.cfg` first.
> 
> I guess it would make sense to delete that in the deployment playbook

Indeed. @assumptionsandg do you think you can make this change?

File: etc/kayobe/ansible/deploy-os-capacity-exporter.yml
Comment (priteau) [inline etc/kayobe/ansible/deploy-os-capacity-exporter.yml]: This should be delegated to the hosts running HAProxy, which are network nodes by default.

**Review by priteau — CHANGES_REQUESTED** [review]

File: etc/kayobe/ansible/deploy-os-capacity-exporter.yml
Comment (holliefae) [reply etc/kayobe/ansible/deploy-os-capacity-exporter.yml]: Changed to `network`

**Review by priteau — CHANGES_REQUESTED** [review]
The "CPUs Used by Project" panel is reporting its values with MB unit.

**Comment by assumptionsandg** [comment]
> The "CPUs Used by Project" panel is reporting its values with MB unit.

Changed unit type to `none`

**Review by priteau — CHANGES_REQUESTED** [review]
I can't see it in the diff but I also spotted a typo `Avaliable` in the dashboard.

**Review by priteau — DISMISSED** [review]

File: etc/kayobe/stackhpc-monitoring.yml
Comment (markgoddard) [inline etc/kayobe/stackhpc-monitoring.yml]: `public_openrc_auth_url` is only defined in the context of the `ansible/public-openrc.yml` playbook in kayobe, and would not be valid here.

**Review by markgoddard — CHANGES_REQUESTED** [review]

File: etc/kayobe/stackhpc-monitoring.yml
Comment (markgoddard) [inline etc/kayobe/stackhpc-monitoring.yml]: internal_net is a specific network that is not guaranteed to be defined. You'd need to use the generic `internal_net_name | net_vip_address` formulation. But it will be easier to use `kolla_internal_fqdn`. In general it's best to use the FQDN rather than the VIP address, since this is what the certificate will be valid for.

```suggestion
stackhpc_os_capacity_auth_url: "http{% if kolla_enable_tls_internal | bool %}s{% endif %}://{{ kolla_internal_fqdn }}:5000"
```

File: doc/source/configuration/monitoring.rst
Comment (markgoddard) [inline doc/source/configuration/monitoring.rst]: Please update to match the proposed change

File: doc/source/configuration/monitoring.rst
Comment (markgoddard) [inline doc/source/configuration/monitoring.rst]: It's a URL, not just the VIP address/FQDN required here

File: etc/kayobe/stackhpc-monitoring.yml
Comment (holliefae) [reply etc/kayobe/stackhpc-monitoring.yml]: Replaced with `internal_vip_address`. It's not elegant but hopefully will make an ok default until #829 is merged.

File: releasenotes/notes/os-capacity-94006f03f16583e4.yaml
Comment (markgoddard) [inline releasenotes/notes/os-capacity-94006f03f16583e4.yaml]: As above

**Review by markgoddard — CHANGES_REQUESTED** [review]

**Review by markgoddard — APPROVED** [review]
Let's get this thing in.

**Pull request merged**
